src/internal/TokenRegistry.ts

```diff
@@ -0,0 +1,78 @@
+import { Tokens } from ".";
+
+type TokenMetadata = {
+  address: string;
+  symbol: string;
+  decimals: number;
+  chainId: number;
+};
+
+type TokenDetails = {
+  address: string;
+  decimals: number;
+};
+
+export default class TokenRegistry {
+  private static instance: TokenRegistry;
+
+  private addressMap: Map<string, TokenMetadata>;
+  private symbolMap: Map<string, TokenMetadata>;
+
+  constructor(tokens: TokenMetadata[]) {
+    this.addressMap = new Map();
+    this.symbolMap = new Map();
+
+    for (const token of tokens) {
+      const normalizedAddress = this.createKey(token.address, token.chainId);
+      const normalizedSymbol = this.createKey(token.symbol, token.chainId);
+
+      this.addressMap.set(normalizedAddress, token);
+      this.symbolMap.set(normalizedSymbol, token);
+    }
+  }
+
+  static getInstance(): TokenRegistry {
+    if (!TokenRegistry.instance) {
+      const tokens: TokenMetadata[] = Tokens.tokens
+      TokenRegistry.instance = new TokenRegistry(tokens);
+    }
+    return TokenRegistry.instance;
+  }
+
+  private normalize(value: string): string {
+    return value.toLowerCase().trim();
+  }
+
+  private createKey(value: string, chainId: number): string {
+    return `${this.normalize(value)}-${chainId}`;
+  }
+
+  getByAddressOrSymbol(addressOrSymBol: string, chainId: number): TokenDetails {
+    const key = this.createKey(addressOrSymBol, chainId);
+    const token = this.addressMap.get(key) || this.symbolMap.get(key);
+
+    if (!token) throw new Error(`Token not found for input "${addressOrSymBol}". Ensure the input is a valid address or symbol.`);
+
+    return {
+      address: token.address,
+      decimals: token.decimals
+    };
+  }
+
+  getByTokenArgs({ address, decimals }: { address: string; decimals?: number }, chainId: number): TokenDetails {
```

Comment:
very confusing fn sig, should just be single object `{ address, chainId, decimals? }`

Reply:
yeah because address and decimals are from TokenArgs. 

so caller can do `geByTokenArgs(tokenArgs, chainId)` wdyt?